### ISSUE TITLE
Float32 precision loss causing cross_entropy loss to return incorrect results

### ISSUE THREAD

**Issue by cui-83** [tensorflow/tensorflow#62204]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.8

### Custom code

Yes

### Current behavior?

```
import tensorflow.keras.backend as K
K.binary_crossentropy([[0.0],[1.0]],[[1.0],[0.0]], from_logits=False)
<tf.Tensor: shape=(2, 1), dtype=float32, numpy=
array([[15.333239],
       [15.424949]], dtype=float32)>
```
In theory, crossentropy isn't symmetric but the non-trivial discrepancy between crossentropy(0,1) and crossentropy(1,0) here is caused by precision loss. Specifically, 
In https://github.com/tensorflow/tensorflow/blob/v2.3.0/tensorflow/python/keras/backend.py#L4825C67-L4825C67
```
epsilon_ = _constant_to_tensor(epsilon(), output.dtype.base_dtype)
output = clip_ops.clip_by_value(output, epsilon_, 1. - epsilon_)

bce = target * math_ops.log(output + epsilon())
bce += (1 - target) * math_ops.log(1 - output + epsilon())
return -bce
```
The problem is, when output==1, 1 - output + epsilon() = 1 - (1. - epsilon_) + epsilon() = 2* epsilon_
the expected result should be 2e-7, but the actual result is 2.1920928e-07 due to precision loss. After taking the natural log, we get 15.333239 and 15.424949. A 0.092 difference in a single sample loss is pretty significant.

### Standalone code to reproduce the issue

```shell
import tensorflow.keras.backend as K
K.binary_crossentropy([[0.0],[1.0]],[[1.0],[0.0]], from_logits=False)
```

### Relevant log output

```shell
<tf.Tensor: shape=(2, 1), dtype=float32, numpy=array([[15.333239],  [15.424949]], dtype=float32)>
```

**Comment by maksym33** [comment]
@cui-83 can you try calculating the loss directly from logits? 
From what I understand this is advised to do exactly for this reason.

**Comment by sushreebarsa** [comment]
Hello, @cui-83!
Sorry for the late response!
Float32 precision has a limited range of re-presentable values, and when dealing with large numbers, the loss calculations can overflow or underflow, resulting in inaccurate values which are intended. Could you please try to use higher precision data types like Float64 or BFloat16 for the loss calculations to avoid such issue?
Thank you!
